File: clients/ts/src/pipes.ts
import { err, ok } from "./errors.js";
import { request, tenantParam } from "./http.js";
import type { StreamController } from "./stream/controller.js";
import type {
  HttpContext,
  OpsRequestOptions,
  Pipe,
  PipeRequestOptions,
  Result,
  StreamOptions,
} from "./types.js";

type CreateStreamFn<Row> = (table: string, opts?: StreamOptions) => StreamController<Row>;

/** Reference to a named query pipe — PromiseLike for convenient `await`. */
export class PipeRef<Row = Record<string, unknown>> implements PromiseLike<Result<Row[]>> {
  private readonly _ctx: HttpContext;
  private readonly _name: string;
  private readonly _params?: Record<string, unknown>;
  private readonly _createStream: CreateStreamFn<Row>;

  constructor(
    ctx: HttpContext,
    name: string,
    params: Record<string, unknown> | undefined,
    createStream: CreateStreamFn<Row>,
  ) {
    this._ctx = ctx;
    this._name = name;
    this._params = params;
    this._createStream = createStream;
  }

  /**
   * Execute the pipe and return results.
   *
   * Takes only `signal` — deliberately narrower than the `RequestOptions` the
   * query builder accepts. The pipes endpoint binds the body as the pipe's
   * parameters, so there is no row cap to forward; give the pipe a `{{limit}}`
   * parameter in its SQL and pass it via `wh.pipe(name, { limit })`.
   */
  async fetch(opts?: PipeRequestOptions): Promise<Result<Row[]>> {
    const { data, error } = await request<Row[]>(this._ctx, {
      method: "POST",
      path: `/v1/pipes/${encodeURIComponent(this._name)}`,
      body: this._params ?? {},
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }

  /** Subscribe to live events from the pipe's underlying query. */
  stream(opts?: StreamOptions): StreamController<Row> {
    return this._createStream(this._name, opts);
  }

  then<TResult1 = Result<Row[]>, TResult2 = never>(
    onfulfilled?: ((value: Result<Row[]>) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null,
  ): Promise<TResult1 | TResult2> {
    return this.fetch().then(onfulfilled, onrejected);
  }
}

/**
 * Admin namespace for reading named query pipes.
 *
 * Pipes are defined in the server's settings directory (`pipes.json`, with
 * every `allowed_roles` entry declared in `roles.json`); the files are the only
 * way to define or change a pipe. Edit the files and the server re-adopts them on change, on
 * SIGHUP, or on `wh.settings.reload()` (POST /v1/ops/settings/reload).
 */
export class PipesNamespace {
  private readonly _ctx: HttpContext;

  constructor(ctx: HttpContext) {
    this._ctx = ctx;
  }

  /** List all registered pipes — of `opts.tenant`, the default tenant without it. */
  async list(opts?: OpsRequestOptions): Promise<Result<Pipe[]>> {
    const { data, error } = await request<Pipe[]>(this._ctx, {
      method: "GET",
      path: "/v1/ops/pipes",
      params: tenantParam(opts),
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }

  /** Get a single pipe definition by name — of `opts.tenant`, the default tenant without it. */
  async get(name: string, opts?: OpsRequestOptions): Promise<Result<Pipe>> {
    const { data, error } = await request<Pipe>(this._ctx, {
      method: "GET",
      path: `/v1/ops/pipes/${encodeURIComponent(name)}`,
      params: tenantParam(opts),
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }
}

File: clients/go/pipes.go
package wavehouse

import (
	"context"
	"fmt"
	"net/url"
)

// PipesNamespace provides admin-only named-pipe management.
type PipesNamespace struct {
	ctx httpContext
}

// List returns all registered pipes. Admin-only.
func (p *PipesNamespace) List(ctx context.Context) ([]Pipe, error) {
	var pipes []Pipe
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "GET",
		path:   "/v1/ops/pipes",
	}, &pipes); err != nil {
		return nil, fmt.Errorf("list pipes: %w", err)
	}
	return pipes, nil
}

// Get returns a single pipe definition by name. Admin-only.
func (p *PipesNamespace) Get(ctx context.Context, name string) (*Pipe, error) {
	var pipe Pipe
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "GET",
		path:   "/v1/ops/pipes/" + url.PathEscape(name),
	}, &pipe); err != nil {
		return nil, fmt.Errorf("get pipe %q: %w", name, err)
	}
	return &pipe, nil
}

// Set creates or updates a pipe. Admin-only.
func (p *PipesNamespace) Set(ctx context.Context, name string, def PipeDef) error {
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "PUT",
		path:   "/v1/ops/pipes/" + url.PathEscape(name),
		body:   def,
	}, nil); err != nil {
		return fmt.Errorf("set pipe %q: %w", name, err)
	}
	return nil
}

// Delete removes a pipe by name. Admin-only.
func (p *PipesNamespace) Delete(ctx context.Context, name string) error {
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "DELETE",
		path:   "/v1/ops/pipes/" + url.PathEscape(name),
	}, nil); err != nil {
		return fmt.Errorf("delete pipe %q: %w", name, err)
	}
	return nil
}

// PipeDef is the definition body for creating/updating a pipe (Pipe minus name).
type PipeDef struct {
	SQL          string     `json:"sql"`
	Parameters   []ParamDef `json:"parameters,omitempty"`
	Description  string     `json:"description,omitempty"`
	AllowedRoles []string   `json:"allowed_roles,omitempty"`
}

// PipeRef is a reference to a named query pipe. Use Fetch to execute it.
type PipeRef struct {
	ctx          httpContext
	name         string
	params       map[string]any
	createStream func(table string, opts *StreamOptions) *StreamController
}

// Fetch executes the pipe and returns the result rows decoded into []T.
func Fetch[Row any](ctx context.Context, p *PipeRef) ([]Row, error) {
	body := p.params
	if body == nil {
		body = map[string]any{}
	}
	var rows []Row
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "POST",
		path:   "/v1/pipes/" + url.PathEscape(p.name),
		body:   body,
	}, &rows); err != nil {
		return nil, fmt.Errorf("execute pipe %q: %w", p.name, err)
	}
	return rows, nil
}

// FetchUntyped executes the pipe and returns rows as []map[string]any.
func (p *PipeRef) FetchUntyped(ctx context.Context) ([]map[string]any, error) {
	return Fetch[map[string]any](ctx, p)
}

// Stream subscribes to live events using the pipe's name as a table name. The
// pipe's SQL and params are NOT applied: where a table of that name exists the
// caller receives its raw events, and otherwise the stream stays silent. Kept
// for parity with the TypeScript SDK's PipeRef.stream(), which behaves the same
// way; both wait on a pipe-aware stream endpoint (issue #445).
func (p *PipeRef) Stream(opts *StreamOptions) *StreamController {
	return p.createStream(p.name, opts)
}
